jaxrs/src/main/java/com/facebook/airlift/jaxrs/thrift/ThriftMapper.java

```diff
@@ -0,0 +1,122 @@
+
+    private final ThriftCodecManager thriftCodecManager;
+
+    @Inject
+    public ThriftMapper(ThriftCodecManager thriftCodecManager)
+    {
+        this.thriftCodecManager = requireNonNull(thriftCodecManager, "thriftCodecManager is null");
+    }
+
+    @Override
+    public Object readFrom(
+            Class<Object> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, String> httpHeaders,
+            InputStream inputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec<?> codec = thriftCodecManager.getCodec(genericType);
+            Object value = ThriftProtocolUtils.read(codec, getThriftProtocol(mediaType), inputStream);
+            return value;
+        }
+        catch (Exception e) {
+            // we want to return a 400 for bad Thrift but not for a real IO exception
+            if (e instanceof IOException && !(e instanceof EOFException)) {
+                throw (IOException) e;
+            }
+            //The IOException is likely to be wrapped into a TTransportException
+            if (e instanceof TTransportException && e.getCause() instanceof IOException && !(e.getCause() instanceof EOFException)) {
+                throw (IOException) e.getCause();
+            }
+            // log the exception at debug so it can be viewed during development
+            // Note: we are not logging at a higher level because this could cause a denial of service
+            log.debug(e, "Invalid Thrift input for Java type %s", genericType);
+            // Invalid thrift request. Throwing exception so the response code can be overridden using a mapper.
+            throw new ThriftMapperParsingException(type, e);
+        }
+    }
+
+    @Override
+    public void writeTo(
+            Object value,
+            Class<?> type,
+            Type genericType,
+            Annotation[] annotations,
+            MediaType mediaType,
+            MultivaluedMap<String, Object> httpHeaders,
+            OutputStream outputStream)
+            throws IOException
+    {
+        try {
+            ThriftCodec codec = thriftCodecManager.getCodec(type);
+            ThriftProtocolUtils.write(value, codec, getThriftProtocol(mediaType), outputStream);
+        }
+        catch (Exception e) {
+            // handing EOFException same as JsonMapper
+            if (e instanceof EOFException || (e instanceof TTransportException && e.getCause() instanceof EOFException)) {
+                return;
+            }
+            log.debug(e, "Can not serialize to thrift for Java type %s", type);
+            throw new ThriftMapperParsingException(type, e);
+        }
+    }
+
+    private ThriftProtocol getThriftProtocol(MediaType mediaType)
+    {
+        ThriftProtocol thriftProtocol = BINARY;
+        if (mediaType != null && mediaType.getParameters().containsKey("t")) {
```

Comment:
Let's not assume the protocol, throw an error instead if the content type / parameter is missing